numpy/_core/src/multiarray/unique.cpp

```diff
@@ -32,11 +35,116 @@ FinalAction<F> finally(F f) {
 }
 
 template <typename T>
+inline int equal_default(T lhs, T rhs) {
+    return lhs == rhs;
+}
+
+// As npy_isnan is a macro, we need to define a wrapper function to use it in a template.
+template <typename T>
+inline int npy_isnan_wrapper(T value) {
+    return npy_isnan(value);
+}
+
+template <typename T>
+inline size_t hash_nonan(const T *value, npy_bool equal_nan) {
+    return std::hash<T>{}(*value);
+}
+
+template <
+    typename T,
+    int (*isnan)(T) = npy_isnan_wrapper<T>
+>
+size_t hash_maybenan(const T *value, npy_bool equal_nan) {
+    if (equal_nan && isnan(*value)) {
+        return 0;
+    }
+    return hash_nonan(value, equal_nan);
+}
+
+template <typename S, typename T, S (*real)(T), S (*imag)(T)>
+size_t hash_complex(const T *value, npy_bool equal_nan) {
+    S value_real = real(*value);
+    S value_imag = imag(*value);
+    int isnan = npy_isnan_wrapper<S>(value_real) || npy_isnan_wrapper<S>(value_imag);
+    if (equal_nan && isnan) {
+        return 0;
+    }
+
+    return hash_nonan<S>(&value_real, equal_nan) ^ hash_nonan<S>(&value_imag, equal_nan);
```

Comment:
Mentioned at the triage meeting: using xor for hash combine might lead to some hash collisions. Since NumPy already uses fmv-hash, could we instead feed the raw bytes into this function instead? This retain the hash integrity of hash.